cinema/views.py

```diff
@@ -56,3 +83,75 @@
             return MovieSessionDetailSerializer
 
         return MovieSessionSerializer
+
+    def get_queryset(self):
+        queryset = MovieSession.objects.prefetch_related("movie", "cinema_hall", "tickets")
+        date = self.request.query_params.get("date")
+        movie = self.request.query_params.get("movie")
+
+        if date:
+            queryset = queryset.filter(show_time__date=date)
+
+        if movie:
+            queryset = queryset.filter(movie__id=movie)
+
+        if self.action == "list":
+            queryset = queryset.select_related(
+                "movie",
+                "cinema_hall"
+            ).annotate(
+                tickets_available=F(
+                    "cinema_hall__rows"
+                ) * F(
+                    "cinema_hall__seats_in_row"
+                ) - Count(
+                    "tickets"
+                )
+            )
+
+        elif self.action == "retrieve":
+            queryset = queryset.prefetch_related("movie", "cinema_hall")
+
+        return queryset.distinct()
```

Comment:
The `get_queryset` method in `MovieSessionViewSet` correctly filters sessions based on date and movie. It also calculates `tickets_available` using annotations, which is a good practice for efficient database querying.